### PR TITLE
#[UIC-719] URL redirection changes with url_for flask method

### PR DESCRIPTION
**Pull request by ankursinghal2005**
[UIC-719](https://guavus-jira.atlassian.net/browse/UIC-719)

### CATEGORY

Choose one

- [ ] Bug Fix
- [x] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
URL redirection changes with url_for flask method

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS
@bipinsoniguavus 
@jitendra-kumawat 
@arpit-agarwal

### CODE REVIEW

File: superset/views/core.py [hunk 1]
```diff
@@ -828,8 +826,7 @@ def shortner(self):
         obj = models.Url(url=url)
         db.session.add(obj)
         db.session.commit()
-        return Response(
-            '/r/{obj.id}'.format(obj=obj),
+        return Response(url_for('R') + '{obj.id}'.format(obj=obj),
```
Comment (ankursinghal2005) [inline superset/views/core.py]: This is not working. We have to replace proper endpoint value in url_for.

File: superset/views/core.py [hunk 2]
```diff
@@ -3035,17 +3037,21 @@ class CssTemplateAsyncModelView(CssTemplateModelView):
...
     icon='fa-search',
     category_icon='fa-flask',
     category='SQL Lab',
     category_label=__('SQL Lab'),
 )
 
+qs_menu_item = appbuilder.menu.find('Query Search')
+qs_get_url = qs_menu_item.get_url
+qs_menu_item.get_url = lambda: qs_get_url() + '#search'
+
```
Comment (jitendra-kumawat) [inline superset/views/core.py]: can we find a way to pass params in url_for method ....instead setting it manually

File: superset/connectors/sqla/views.py [hunk 3]
```diff
@@ -380,17 +380,19 @@ def refresh(self, tables):
...
+        return redirect(url_for('TableModelView.list'))
 
 
 appbuilder.add_view_no_menu(TableModelView)
 appbuilder.add_link(
     'Tables',
     label=__('Tables'),
-    href='/tablemodelview/list/?_flt_1_is_sqllab_view=y',
+    href='TableModelView.list',
     icon='fa-table',
```
[the remark this answers is not in the document]
Comment (bipinsoniguavus) [reply superset/connectors/sqla/views.py]: Verified it with current. It doesn't affect any functionality. We can remove it.

File: superset/models/core.py [hunk 4]
```diff
@@ -420,12 +421,11 @@ def url(self):
                 try:
                     if json.loads(default_filters):
                         filters = parse.quote(default_filters.encode('utf8'))
-                        return '/superset/dashboard/{}/?preselect_filters={}'.format(
-                            self.slug or self.id, filters)
+                        return url_for('Superset.dashboard' , dashboard_id='{}'.format(self.slug or self.id)) + '?preselect_filters={}'.format(filters)
                 except Exception:
```
[the remark this answers is not in the document]
Comment (bipinsoniguavus) [reply superset/models/core.py]: `format()` is one of the string formatting methods in Python, which allows multiple substitutions and value formatting. This method lets us concatenate elements within a string through positional formatting. and superset uses the same and to make it consistent, we should not change the way it is written.

File: superset/templates/superset/fab_overrides/list_with_checkboxes.html [hunk 5]
```diff
@@ -82,7 +82,7 @@
                           {{'checked' if item[value] }}
                           name="{{ '{}__{}'.format(pk, value) }}"
                           id="{{ '{}__{}'.format(pk, value) }}"
-                          data-checkbox-api-prefix="/superset/checkbox/{{ modelview_name }}/{{ pk }}/{{ value }}/">
+                          data-checkbox-api-prefix="{{ url_for('Superset.checkbox', model_view=modeview_name, id_=pk, attr=value, value='') }}">
                     {% else %}
```
[the remark this answers is not in the document]
Comment (bipinsoniguavus) [reply superset/templates/superset/fab_overrides/list_with_checkboxes.html]: It is required to be here, please refer [here](https://github.com/Guavus/incubator-superset/blob/9af22825d91a1684bff6167477717980a812e77c/superset/views/core.py#L1582) for method signature

File: superset/views/core.py (hunk 2 above)
Comment (bipinsoniguavus) [reply superset/views/core.py]: As discussed with @ankursinghal2005 this is how we can handle links with `href`. If we find better resolution in near future, we can adopt the same.

File: superset/views/core.py (hunk 1 above)
Comment (jitendra-kumawat) [reply superset/views/core.py]: replaced with older one.

**Pull request merged**